src/graph/ford_fulkerson.rs

```diff
@@ -1,141 +1,311 @@
-            vec![0, 0, 6, 0, 0, 6],
-            vec![0, 0, 0, 0, 0, 0],
-        ];
-        assert_eq!(ford_fulkerson(&graph, 0, 5), 7);
-    }
-
-    #[test]
-    fn test_example_3() {
-        let graph = vec![
-            vec![0, 10, 0, 10, 0, 0],
-            vec![0, 0, 4, 2, 8, 0],
-            vec![0, 0, 0, 0, 0, 10],
-            vec![0, 0, 0, 0, 9, 0],
-            vec![0, 0, 6, 0, 0, 10],
-            vec![0, 0, 0, 0, 0, 0],
-        ];
-        assert_eq!(ford_fulkerson(&graph, 0, 5), 19);
-    }
-
-    #[test]
-    fn test_example_4() {
-        let graph = vec![
-            vec![0, 8, 0, 0, 3, 0],
-            vec![0, 0, 9, 0, 0, 0],
-            vec![0, 0, 0, 0, 7, 2],
-            vec![0, 0, 0, 0, 0, 5],
-            vec![0, 0, 7, 4, 0, 0],
-            vec![0, 0, 0, 0, 0, 0],
-        ];
-        assert_eq!(ford_fulkerson(&graph, 0, 5), 6);
+    macro_rules! test_max_flow {
+        ($($name:ident: $tc:expr,)* ) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (graph, source, sink, expected_result) = $tc;
+                    let result = ford_fulkerson(&graph, source, sink);
+                    assert_eq!(result, expected_result);
+                }
+            )*
+        };
     }
 
-    #[test]
-    fn test_example_5() {
-        let graph = vec![
-            vec![0, 16, 13, 0, 0, 0],
-            vec![0, 0, 10, 12, 0, 0],
-            vec![0, 4, 0, 0, 14, 0],
-            vec![0, 0, 9, 0, 0, 20],
-            vec![0, 0, 0, 7, 0, 4],
-            vec![0, 0, 0, 0, 0, 0],
-        ];
-        assert_eq!(ford_fulkerson(&graph, 0, 5), 23);
+    test_max_flow! {
+        test_empty_graph: (
+            vec![],
+            0,
+            0,
+            Err(FordFulkersonError::EmptyGraph),
+        ),
+        test_graph_with_empty_edge: (
+            vec![
+                vec![],
+            ],
+            0,
+            5,
+            Err(FordFulkersonError::EmptyGraph),
+        ),
+        test_source_out_of_bound: (
+            vec![
+                vec![0, 8, 0, 0, 3, 0],
+                vec![0, 0, 9, 0, 0, 0],
+                vec![0, 0, 0, 0, 7, 2],
+                vec![0, 0, 0, 0, 0, 5],
+                vec![0, 0, 7, 4, 0, 0],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            6,
+            5,
+            Err(FordFulkersonError::SourceOutOfBounds),
+        ),
+        test_sink_out_of_bound: (
+            vec![
+                vec![0, 8, 0, 0, 3, 0],
+                vec![0, 0, 9, 0, 0, 0],
+                vec![0, 0, 0, 0, 7, 2],
+                vec![0, 0, 0, 0, 0, 5],
+                vec![0, 0, 7, 4, 0, 0],
+                vec![0, 0, 0, 0, 0, 0],
+            ],
+            0,
+            6,
+            Err(FordFulkersonError::SinkOutOfBounds),
+        ),
+        test_improper_graph: (
+            vec![
+                vec![0, 8],
+                vec![0, 0, 9],
+                vec![0, 0, 0, 0, 7],
```

Comment:
Why not to use a _smaller_ test case?

```suggestion
                vec![0, 8],
                vec![0],
```

Reply:
Just use it now.